sentry/src/main/java/io/sentry/ServerNameResolvingEventProcessor.java

```diff
@@ -0,0 +1,134 @@
+   * time defined by {@link #GET_HOSTNAME_TIMEOUT} without result.
+   */
+  static final class HostnameCache {
+    /** Time before the get hostname operation times out (in ms). */
+    private static final long GET_HOSTNAME_TIMEOUT = TimeUnit.SECONDS.toMillis(1);
+    /** Time for which the cache is kept. */
+    private final long cacheDuration;
+    /** Current value for hostname (might change over time). */
+    @Nullable private volatile String hostname;
+    /** Time at which the cache should expire. */
+    private volatile long expirationTimestamp;
+    /** Whether a cache update thread is currently running or not. */
+    private final @NotNull AtomicBoolean updateRunning = new AtomicBoolean(false);
+
+    private final @NotNull Callable<InetAddress> getLocalhost;
+
+    private HostnameCache(long cacheDuration) {
+      this(cacheDuration, InetAddress::getLocalHost);
+    }
+
+    /**
+     * Sets up a cache for the hostname.
+     *
+     * @param cacheDuration cache duration in milliseconds.
+     * @param getLocalhost a callback to obtain the localhost address - this is mostly here because
+     *     of testability
+     */
+    HostnameCache(long cacheDuration, final @NotNull Callable<InetAddress> getLocalhost) {
+      this.cacheDuration = cacheDuration;
+      this.getLocalhost = Objects.requireNonNull(getLocalhost, "getLocalhost is required");
+    }
+
+    /**
+     * Gets the hostname of the current machine.
+     *
+     * <p>Gets the value from the cache if possible otherwise calls {@link #updateCache()}.
+     *
+     * @return the hostname of the current machine.
+     */
+    private @Nullable String getHostname() {
+      if (expirationTimestamp < System.currentTimeMillis()
+          && updateRunning.compareAndSet(false, true)) {
+        updateCache();
+      }
+
+      return hostname;
+    }
+
+    /** Force an update of the cache to get the current value of the hostname. */
+    private void updateCache() {
+      final Callable<Void> hostRetriever =
+          () -> {
+            try {
+              hostname = getLocalhost.call().getCanonicalHostName();
+              expirationTimestamp = System.currentTimeMillis() + cacheDuration;
+            } finally {
+              updateRunning.set(false);
+            }
+
+            return null;
+          };
+
+      try {
+        final FutureTask<Void> futureTask = new FutureTask<>(hostRetriever);
+        new Thread(futureTask).start();
+        futureTask.get(GET_HOSTNAME_TIMEOUT, TimeUnit.MILLISECONDS);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        handleCacheUpdateFailure();
+      } catch (ExecutionException | TimeoutException | RuntimeException e) {
+        handleCacheUpdateFailure();
+      }
+    }
+
+    private void handleCacheUpdateFailure() {
+      expirationTimestamp = System.currentTimeMillis() + TimeUnit.SECONDS.toMillis(1);
```

Comment:
there's a class called `CurrentDateProvider` which returns a `System.currentTimeMillis()` so you could unit test if necessary.

Reply:
If it fails to resolve, it'll kick off again after 1 second.
We catch any error here and set it to 1 second again. So basically this can mean we keep spawning a new thread every second. I think in case of error we should just stop running this and just return `null`.

WHat makes this fail? Are those errors recoverable? is this a reverse DNS query on the local interface IP addresses?

Reply:
Failure is unlikely to happen, I can't think of the case when it would actually happen. Code has changed to use single daemon thread so that at least we don't spawn new threads all the time.